### PR TITLE
Adding download_all task to poeditor

### PR DESCRIPTION
**Pull request by yosiat**
Adding download_all task for poeditor.
Using this we won't need to update poeditor grunt files when a new language appears.

See README.md for how this works.

Asna: https://app.asana.com/0/207822667006350/228700371960775

### CODE REVIEW

**Review by ghost — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -24,4 +24,31 @@ poeditor: {
...
+poeditor: {
+  download_all: {
+    // Same configuration as download
+    download_all: {
+      dest: 'tests/?.json',
+
+      // Specify minimum coverage here
+      minimum_coverage: 40
+    }
+  },
```
Comment (eliranha) [inline README.md:42]: download_all.download_all ?

File: tasks/poeditor.js [hunk 2]
```diff
@@ -26,15 +23,64 @@ module.exports = function(g) {
...
+            var downloadMeta = confLanguages(data.download, opts);
+            downloadMeta.project = opts.project;
+            download(downloadMeta, opts, this.async());
+        }
+        // download_all
+        else if (data.download_all) {
+            var done = this.async();
+            getLanguagesToDownload(data.download_all, opts, function(err, languages) {
+              if(err) {
+                done(false);
```
Comment (eliranha) [inline tasks/poeditor.js]: done(err)

File: README.md (hunk 1 above)
Comment (yosiat) [reply README.md:42]: the first is the name of the task, the second is real grunt task.
we have `download.download` in our gruntfile under `poeditor`.

File: tasks/poeditor.js [hunk 3]
```diff
@@ -26,15 +23,64 @@ module.exports = function(g) {
...
+                return;
+              }
+
+              opts.languages = _.extend(languages, opts.languages_override || {});
```
Comment (eliranha) [inline tasks/poeditor.js]: _.extend(languages, opts.languages_override || {});
you can skip the || {}
.extend handles it.

File: tasks/poeditor.js [hunk 4]
```diff
@@ -26,15 +23,64 @@ module.exports = function(g) {
...
+    };
+
+    var endpoint = 'https://poeditor.com/api/';
+    request.post(endpoint, {form: creds}, function(err, response, body) {
+        if (err) {
+          callback(err);
+          return;
+        }
+
+        body = JSON.parse(body);
```
Comment (eliranha) [inline tasks/poeditor.js]: use try catch for parsing json, but i think request already support json type

File: tasks/poeditor.js [hunk 5]
```diff
@@ -26,15 +23,64 @@ module.exports = function(g) {
...
+        if(!body.list) {
+          callback('No body.list');
```
Comment (eliranha) [inline tasks/poeditor.js]: callback(new Error('new body.list'));

File: tasks/poeditor.js [hunk 6]
```diff
@@ -26,15 +23,64 @@ module.exports = function(g) {
...
+          return;
+        }
+
+        var languagesToDownload = _.filter(body.list, function(lang) {
+          return lang.percentage > data.minimum_coverage;
         });
-};
+
+        var languagesList = _.reduce(languagesToDownload, function(result, lang) {
```
Comment (eliranha) [inline tasks/poeditor.js]: i think you should avoid _reduce, not the most readable function

File: tasks/poeditor.js [hunk 7]
```diff
@@ -26,15 +23,64 @@ module.exports = function(g) {
...
-                download(downloadMeta, opts, this.async());
-            }
+function getLanguagesToDownload(data, config, callback) {
+    var creds = {
+        api_token: config.apiToken,
+        id: config.project,
+        action: 'list_languages'
+    };
+
+    var endpoint = 'https://poeditor.com/api/';
```
Comment (eliranha) [inline tasks/poeditor.js]: should be defined as const

**Review by eliranha — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (yosiat) [reply README.md:42]: Because you and @roir asked about it, I changed it and clarified it a bit.

Comment (eliranha) [reply README.md:42]: download_languages is not the best name
thats the purpose of poeditor
why not "languages" and nested is "all"

**Comment by eliranha** [comment]
once done, please use your local env to test it (need to update package json)
and deploy from clean environment to staging or app2
also test multi langs

**Comment by yosiat** [comment]
@eliranha Added two features, please review:

- Allowing to pass options under download_all task - useful when you to download for two poeditor projects and you don't want global optinos
- Allowing to disable expands - for tracking app.

File: tasks/poeditor.js [hunk 8]
```diff
@@ -26,15 +25,63 @@ module.exports = function(g) {
...
+        // download
+        else if (data.download) {
+            var downloadMeta = confLanguages(data.download, opts);
+            downloadMeta.project = opts.project;
+            download(downloadMeta, opts, this.async());
+        }
+        // download_all
+        else if (data.download_all) {
+            var done = this.async();
+            opts = data.download_all.options || opts;
```
Comment (eliranha) [inline tasks/poeditor.js]: why not extend ?

Comment (yosiat) [reply tasks/poeditor.js]: changed.

**Comment by eliranha** [comment]
LGTM, lets merge this next week, not today

**Pull request merged**